schema.development.kf

```diff
@@ -1282,8 +1228,56 @@ procedure migrate_all_data($old_dbid text) public owner {
     migrate_shared_user_attrs($old_dbid);
     migrate_inserters($old_dbid);
     migrate_delegates($old_dbid);
-    migrate_external_grants($old_dbid);
-    migrate_write_grants($old_dbid); // WGs must be inserted before internal AGs
-    migrate_internal_grants($old_dbid);
+    migrate_write_grants($old_dbid); // WGs must be inserted before access grants
+    migrate_access_grants($old_dbid);
     migrate_configs($old_dbid);
 }
+
+procedure migrate_ext_grants_as_owner(
+    $id uuid,
+    $ag_owner text,
+    $ag_grantee text,
+    $data_id uuid,
+    $locked_until int,
+    $ag_chain text,
+    $tx_hash text,
+    $block_hash text,
+    $height int,
+    $status text
+) public owner {
+    // We don't need to insert non valid AGs
+    if $status != 'valid' {
+        return;
+    }
+
+    // Deleting credentials or even users, don't trigger removal external AGs from DB. So we need to check the user still exists.
+    $user_id text := '';
+    for $row in SELECT user_id FROM credentials WHERE id = $data_id {
+        $user_id := $row.user_id::text;
+    }
+    if $user_id == '' {
+        return;
+    }
+
+    // There can be copy is exists and shared, but original credential was deleted.
+    // In this case, as original credential id is used for determining AG's is, we will set credential `id` as id of external AG
+    $original_credential_id text := '';
+    for $row2 in SELECT original_id FROM shared_credentials WHERE copy_id = $data_id {
+        $original_credential_id := $row2.original_id::text;
+    }
+
+    $id_for_creation uuid := $id;
+    if $original_credential_id != '' {
+        $id_for_creation := $original_credential_id::uuid;
+    }
```

Comment:
I don't quite understand this tactic. :x

Since deleting the original makes us lose the public notes, how effective does this end up being?

Reply:
This workaround is just to insert existing external AGs for existing copies without original credentials.
if we don't need such AGs we can skip them and do not migrate.

Reply:
I don't see any of these on prod:
```
[ec2-user@alan ~]$ idos-kgw/migrator/distr/kwil-cli database query -n idos "select count(*) from shared_credentials sc where not exists (select 1 from credentials where id = sc.original_id)"
| count |
+-------+
|     0 |
```

Maybe this is not worth the fuss?

Reply:
I think you check the wrong thing.
My case is about:
There were `cred1` as original and `cred2` as copy, record in `shared_credentials`. `ext_ag2` for the copy (`cred2`).
Then `cred1` was deleted, record in `shared_credentials` was deleted also (cascade). So only `cred2` and its `ext_ag2` remain. As we use original credential `id` for `int_ag` id generation, we have to do something in this edge case (no `original_id`, need to substitute by something, not to use `null`).

Reply:
And as it one-time run script, i would not polish it very thoroughly. It is just additional check, no harm from this.